src/jsTypeFactory.cc

```diff
@@ -0,0 +1,175 @@
+/**
+ * @file jsTypeFactory.cc
+ * @author Caleb Aikens (caleb@distributive.network)
+ * @brief
+ * @version 0.1
+ * @date 2023-02-15
+ *
+ * @copyright Copyright (c) 2023
+ *
+ */
+
+#include "include/jsTypeFactory.hh"
+
+#include "include/modules/pythonmonkey/pythonmonkey.hh"
+#include "include/PyType.hh"
+#include "include/FuncType.hh"
+#include "include/pyTypeFactory.hh"
+#include "include/StrType.hh"
+#include "include/IntType.hh"
+
+#include <jsapi.h>
+#include <jsfriendapi.h>
+
+#include <Python.h>
+
+#define HIGH_SURROGATE_START 0xD800
+#define LOW_SURROGATE_START 0xDC00
+#define LOW_SURROGATE_END 0xDFFF
+#define BMP_END 0x10000
+
+struct PythonExternalString : public JSExternalStringCallbacks {
+  void finalize(char16_t *chars) const override {}
+  size_t sizeOfBuffer(const char16_t *chars, mozilla::MallocSizeOf mallocSizeOf) const override {
+    return 0;
+  }
+};
+
+static constexpr PythonExternalString PythonExternalStringCallbacks;
+
+size_t UCS4ToUTF16(const uint32_t *chars, size_t length, uint16_t *outStr) {
+  uint16_t utf16String[length*2];
+  size_t utf16Length = 0;
+
+  for (size_t i = 0; i < length; i++) {
+    if (chars[i] < HIGH_SURROGATE_START || (chars[i] > LOW_SURROGATE_END && chars[i] < BMP_END)) {
+      utf16String[utf16Length] = uint16_t(chars[i]);
+      utf16Length += 1;
+    }
+    else {
+      /* *INDENT-OFF* */
+      utf16String[utf16Length]      = uint16_t(((0b1111'1111'1100'0000'0000 & (chars[i] - BMP_END)) >> 10) + HIGH_SURROGATE_START);
+      utf16String[utf16Length + 1]  = uint16_t(((0b0000'0000'0011'1111'1111 & (chars[i] - BMP_END)) >> 00) +  LOW_SURROGATE_START);
+      utf16Length += 2;
+      /* *INDENT-ON* */
+    }
+  }
+  outStr = (uint16_t *)malloc(sizeof(uint16_t) * utf16Length);
+  memcpy(outStr, utf16String, sizeof(uint16_t) * utf16Length);
+  return utf16Length;
+}
+
+JS::Value jsTypeFactory(JSContext *cx, PyObject *object) {
+  JS::RootedValue returnType(cx);
+
+  if (PyBool_Check(object)) {
+    returnType.setBoolean(PyLong_AsLong(object));
+  }
+  else if (PyLong_Check(object)) {
+    if (PyObject_IsInstance(object, PythonMonkey_BigInt)) { // pm.bigint is a subclass of the builtin int type
+      JS::BigInt *bigint = IntType(object).toJsBigInt(cx);
+      returnType.setBigInt(bigint);
+    } else if (_PyLong_NumBits(object) <= 53) { // num <= JS Number.MAX_SAFE_INTEGER, the mantissa of a float64 is 53 bits (with 52 explicitly stored and the highest bit always being 1)
+      uint64_t num = PyLong_AsLongLong(object);
```

Comment:
```suggestion
      int64_t num = PyLong_AsLongLong(object);
```
This must be signed, otherwise it won't properly handle negative integers

```py
ident = pm.eval("(a) => { return a }")
int1 = -75
print(int1)        // -75
print(ident(int1)) // 1.8446744073709552e+19
print(ident(int1) == int1) // False
```